va/validation-authority.go

```diff
@@ -133,25 +143,55 @@ func (va ValidationAuthorityImpl) validateSimpleHTTP(identifier core.AcmeIdentif
 	}
 
 	httpRequest.Host = hostName
-	tr := &http.Transport{
-		// We are talking to a client that does not yet have a certificate,
-		// so we accept a temporary, invalid one.
-		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
-		// We don't expect to make multiple requests to a client, so close
-		// connection immediately.
-		DisableKeepAlives: true,
-	}
+	tr := *http.DefaultTransport.(*http.Transport)
+	// We are talking to a client that does not yet have a certificate,
+	// so we accept a temporary, invalid one.
+	tr.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
+	// We don't expect to make multiple requests to a client, so close
+	// connection immediately.
+	tr.DisableKeepAlives = true
+	// Intercept Dial in order to connect to the IP address we
+	// selected above.
+	defaultDial := tr.Dial
+	tr.Dial = func(_, _ string) (net.Conn, error) {
+		port := "80"
+		if va.TestMode {
+			port = "5001"
+		} else if scheme == "https" {
+			port = "443"
+		}
+		// Ignore the addr selected by net/http.
+		return defaultDial("tcp", net.JoinHostPort(addrs[0].String(), port))
+	}
+	var redirects []string
+	addrUsed := addrs[0]
 	logRedirect := func(req *http.Request, via []*http.Request) error {
-		va.log.Info(fmt.Sprintf("validateSimpleHTTP [%s] redirect from %q to %q", identifier, via[len(via)-1].URL.String(), req.URL.String()))
+		redirects = append(redirects, req.URL.String())
+		addrs, problem := va.getAddrs(req.URL.Host)
+		if problem != nil {
+			return problem
+		}
+		addrUsed = addrs[0]
+		va.log.Info(fmt.Sprintf("validateSimpleHTTP [%s] redirect from %q to %q [%s]", identifier, via[len(via)-1].URL.String(), req.URL.String(), addrs[0]))
+		challenge.ResolvedAddrs = append(challenge.ResolvedAddrs, addrs...)
```

Comment:
Instead of listing all of the possible addresses we might have used, how about we have `Challenge.ResolvedAddrs` and `Challenge.Redirects` be in 1-1 correspondence, where the address corresponding to a URI is the one we actually used.